diff --git a/dwl.c b/dwl.c
index e32e58f..a7c2e28 100644
--- a/dwl.c
+++ b/dwl.c
@@ -211,6 +211,7 @@ typedef struct {
 	const Layout *lt;
 	enum wl_output_transform rr;
 	int x, y;
+	unsigned tagset;
 } MonitorRule;
 
 struct pointer_constraint {
@@ -997,7 +998,6 @@ createmon(struct wl_listener *listener, void *data)
 	m->gappiv = gappiv;
 	m->gappoh = gappoh;
 	m->gappov = gappov;
-	m->tagset[0] = m->tagset[1] = (1<<getunusedtag()) & TAGMASK;
 	for (r = monrules; r < END(monrules); r++) {
 		if (!r->name || strstr(wlr_output->name, r->name)) {
 			m->mfact = r->mfact;
@@ -1008,6 +1008,7 @@ createmon(struct wl_listener *listener, void *data)
 			wlr_output_set_transform(wlr_output, r->rr);
 			m->m.x = r->x;
 			m->m.y = r->y;
+			m->tagset[0] = m->tagset[1] = (r->tagset ? r->tagset : (1<<getunusedtag())) & TAGMASK;
 			break;
 		}
 	}
