diff --git a/dwl.c b/dwl.c
index 7b1ca6f..242bd2a 100644
--- a/dwl.c
+++ b/dwl.c
@@ -2273,9 +2273,8 @@ setlayout(const Arg *arg)
 {
 	if (!selmon)
 		return;
-	if (!arg || !arg->v || arg->v != selmon->lt[selmon->sellt])
-		selmon->sellt ^= 1;
-	if (arg && arg->v)
+	selmon->sellt ^= 1;
+	if (arg && arg->v && arg->v != selmon->lt[selmon->sellt ^ 1])
 		selmon->lt[selmon->sellt] = (Layout *)arg->v;
 	strncpy(selmon->ltsymbol, selmon->lt[selmon->sellt]->symbol, LENGTH(selmon->ltsymbol));
 	if (!selmon->lt[selmon->sellt]->arrange) {
